feat: add curve twap price feed and constant price feed

Van0k opened the pull request.

Van0k committed:
feat: add curve twap price feed and constant price feed

lekhovitsky commented on contracts/oracles/ConstantPriceFeed.sol:
```diff
@@ -0,0 +1,57 @@
+        price = _price;
+
+        string memory tokenSymbol = IERC20Metadata(_token).symbol();
+        description = string.concat(tokenSymbol, " / USD constant price feed");
```
There's a decent chance of this price feed being used not only for pricing against USD, so maybe the approach with description tickers we use in other places would be more appropriate:

https://github.com/Gearbox-protocol/oracles-v3/blob/fc8d3a0ab5bd7eb50ce3f6b87dde5cd3d887bafe/contracts/oracles/BoundedPriceFeed.sol#L53-L55

lekhovitsky replied:
Thinking more about it, this price feed might not even be token-specific, so something like "1 USD constant price feed" (for hard-coded stables) or "0.9 ETH constant price feed" (for, safe LRT pricing) would make perfect sense

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,109 @@
+
+    /// @notice Returns USD price of the token token with 8 decimals
+    function latestRoundData() external view override returns (uint80, int256 answer, uint256, uint256, uint80) {
+        uint256 exchangeRate = ICurvePool(pool).price_oracle();
```
Is it guaranteed to have 18 decimals?

Van0k replied:
Seems so, even for pools with 6 decimals assets.

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,109 @@
+        upperBound = _upperBound;
+
+        string memory tokenSymbol = IERC20Metadata(_token).symbol();
+        description = string.concat(tokenSymbol, " / USD Curve TWAP price feed");
```
Again, maybe the description ticker would be more informative (say, if we create a composite price feed `yetAnotherUSD / USDC Curve TWAP price feed` times `USDC / USD Chainlink price feed`)

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,109 @@
+
+    /// @notice Serialized price feed parameters
+    function serialize() external view override returns (bytes memory) {
+        return abi.encode(token, pool, priceFeed, lowerBound, upperBound);
```
We can remove `priceFeed` from here I guess:
- serialization usually includes only "non-standard" fields (and stuff like `priceFeed` and `stalenessPeirod` is handled by `NestedPriceFeeds` lib in compressors)
- if we do, we get the matching signature to `LPPriceFeed`'s serialization:

https://github.com/Gearbox-protocol/oracles-v3/blob/fc8d3a0ab5bd7eb50ce3f6b87dde5cd3d887bafe/contracts/oracles/LPPriceFeed.sol#L66-L70

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,109 @@
+        address _priceFeed,
+        uint32 _stalenessPeriod
+    ) nonZeroAddress(_token) nonZeroAddress(_pool) nonZeroAddress(_priceFeed) {
+        if (_upperBound < _lowerBound) revert UpperBoundTooLowException();
```
Should we check for bounds to be non-zero? Or at least force the upper bound to be strictly greater than lower bound.

Van0k replied:
Might want to set lowerBound == upperBound to keep a static value in some cases?

lekhovitsky replied:
Ah, my favourite type, `ConditionalConstantPriceFeed` (reverts if exchange rate falls below `price`, returns `price` otherwise)

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,109 @@
+/// @notice Computes price of coin 1 in a Curve pool in terms of units of coin 0, based on the pool's TWAP, which is then
+///         multiplied by the price of coin 0.
+contract CurveTWAPPriceFeed is IPriceFeed, PriceFeedValidationTrait, SanityCheckTrait {
+    using Address for address;
```
probably redundant

Van0k committed:
fix: minor constant price feed + curve twap improvements

lekhovitsky commented on contracts/oracles/ConstantPriceFeed.sol:
```diff
@@ -0,0 +1,64 @@
+    using LibString for bytes32;
+    /// @notice Contract version
+
+    uint256 public constant override version = 3_10;
```
Cursor breaks formatting again

lekhovitsky commented on contracts/oracles/ConstantPriceFeed.sol:
```diff
@@ -0,0 +1,64 @@
+    bool public constant override skipPriceCheck = true;
+
+    /// @notice The token address this price feed is for
+    address public immutable token;
```
Do we still need to store token?

Van0k replied:
Don't strictly need, might be good for informational purposes.

lekhovitsky replied:
This argument still applies: https://github.com/Gearbox-protocol/oracles-v3/pull/60#discussion_r2090928007

We might set the same "1.00 USD constant price feed" for multiple `yetAnotherUSD` tokens, so the `token` field might even be misleading.

lekhovitsky commented on contracts/oracles/ConstantPriceFeed.sol:
```diff
@@ -0,0 +1,64 @@
+import {LibString} from "@solady/utils/LibString.sol";
+import {IPriceFeed} from "@gearbox-protocol/core-v3/contracts/interfaces/base/IPriceFeed.sol";
+import {SanityCheckTrait} from "@gearbox-protocol/core-v3/contracts/traits/SanityCheckTrait.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
```
Can be removed (as long as we remove `token`)

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,142 @@
+import {WAD} from "@gearbox-protocol/core-v3/contracts/libraries/Constants.sol";
+import {PriceFeedValidationTrait} from "@gearbox-protocol/core-v3/contracts/traits/PriceFeedValidationTrait.sol";
+import {SanityCheckTrait} from "@gearbox-protocol/core-v3/contracts/traits/SanityCheckTrait.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
```
Can remove

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,142 @@
+    ///      Supports TwoCryptoOptimized and CurveStableNG pools
+    ///      Returns 1 / rate if the requested rate is coin1/coin0
+    function _getExchangeRate() internal view returns (uint256 rate) {
+        try ICurvePool(pool).price_oracle() returns (uint256 _rate) {
+            rate = _rate;
+        } catch {
+            try ICurvePool(pool).price_oracle(0) returns (uint256 _rate) {
+                rate = _rate;
+            } catch {
+                revert UnknownCurveOracleSignatureException();
+            }
+        }
```
Is `try-catch` "safe" in Curve? Meaning state-changing callbacks, especially for some weird pools with ETH.

Van0k replied:
This works for the Curve adapter, so should work here.

Van0k replied:
Works for Curve adapters (handling of int128/uint256 signature for view functions are handled with try/catch), so should work here as well.

lekhovitsky commented on contracts/oracles/curve/CurveTWAPPriceFeed.sol:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /// @dev Retrieves the price oracle value from the curve pool
+    ///      Supports TwoCryptoOptimized and CurveStableNG pools
```
Maybe move this line to contract's NatSpec? To have better chances of not missing it.

lekhovitsky replied:
What about other kinds of pools though?
- [crvUSD/USDC](https://etherscan.io/address/0x4dece678ceceb27446b35c672dc7d61f30bad69e) is neither NG nor Crypto
- [crvUSD/ETH/CRV](https://etherscan.io/address/0x4ebdf703948ddcea3b11f675b4d1fba9d2414a14) is TriCryptoOptimized

lekhovitsky replied:
Is it possible we want to use `price_oracle(1)` or even one of `price_oracle(1) / price_oracle(0)` and `price_oracle(0) / price_oracle(1)` in 3-coin pools?

Van0k replied:
Should work crvUSD/USDC (although not sure how to call this category in general - new StableSwap?). Thought about also supporting 3-coin pools, decided against it for the sake of simplicity - there aren't many of them, and they usually have assets with established feeds anyway.

Van0k committed:
fix: minor fixes

lekhovitsky committed:
test: fix tests, remove compiler warnings

Things to take into account:
- need to be careful with order of deployed price feeds
- `evm_version` is set to `cancun` for tests to work but in reality we compile with `shanghai`
